presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/Extensions.kt

```diff
@@ -0,0 +1,72 @@
+package com.exallium.mvvmcleanexample.presentation.utils
+
+import android.databinding.BaseObservable
+import android.databinding.ViewDataBinding
+import com.exallium.mvvmcleanexample.presentation.utils.delegates.DistinctObservableProperty
+import com.exallium.mvvmcleanexample.presentation.utils.delegates.EcoBindingProperty
+import com.exallium.mvvmcleanexample.presentation.utils.delegates.EcoDisposableProperty
+import com.exallium.mvvmcleanexample.presentation.utils.delegates.RelayProperty
+import com.exallium.mvvmcleanexample.presentation.utils.notifiable.NotifiableObservable
+import com.jakewharton.rxrelay2.BehaviorRelay
+import com.jakewharton.rxrelay2.PublishRelay
+import io.reactivex.disposables.Disposable
+import kotlin.properties.ReadWriteProperty
+import kotlin.reflect.KProperty
+
+typealias OnChange<T> = (T) -> Unit
+
+fun <T> distinctObservable(initialValue: T, onChange: OnChange<T>? = null):
+        ReadWriteProperty<Any?, T> = object : DistinctObservableProperty<T>(initialValue) {
+    override fun afterChange(property: KProperty<*>, oldValue: T, newValue: T) {
+        onChange?.invoke(newValue)
+    }
+}
+
+fun <T> BaseObservable.bindable(initialValue: T, bindableResourceId: Int): ReadWriteProperty<Any?, T> {
+    return object : DistinctObservableProperty<T>(initialValue) {
+        override fun afterChange(property: KProperty<*>, oldValue: T, newValue: T) {
+            notifyPropertyChanged(bindableResourceId)
+        }
+    }
+}
+
+fun <T> BaseObservable.bindable(initialValue: T): ReadWriteProperty<Any?, T> {
+    return object : DistinctObservableProperty<T>(initialValue) {
+        override fun afterChange(property: KProperty<*>, oldValue: T, newValue: T) {
+            notifyPropertyChanged(PropertyCacher.bindableResourceIdForKProperty(property))
+        }
+    }
+}
+
+fun <T> NotifiableObservable.bindable(initialValue: T, onChange: OnChange<T>? = null): ReadWriteProperty<Any?, T> {
+    return object : DistinctObservableProperty<T>(initialValue) {
+        override fun afterChange(property: KProperty<*>, oldValue: T, newValue: T) {
+            notifyPropertyChanged(PropertyCacher.bindableResourceIdForKProperty(property))
+            onChange?.invoke(newValue)
+        }
+    }
+}
+
+fun ecoDisposable(disposable: Disposable?): EcoDisposableProperty = EcoDisposableProperty(disposable)
+
+fun <T> publishRelay(initialValue: T? = null): RelayProperty<T> {
```

Comment:
let's document all of these w. /** */ since we're starting to get into more advanced kotlin features (those that don't exist in Java) so that people not so familiar with kotlin can still follow along.